junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/AutoCloseTests.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright 2015-2023 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * https://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.jupiter.engine.extension;
+
+import static java.util.Arrays.asList;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.extension.ExtensionContext.Namespace;
+import static org.junit.platform.testkit.engine.EventConditions.finishedWithFailure;
+import static org.junit.platform.testkit.engine.TestExecutionResultConditions.cause;
+import static org.junit.platform.testkit.engine.TestExecutionResultConditions.message;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.junit.jupiter.api.AutoClose;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.engine.AbstractJupiterTestEngineTests;
+import org.junit.platform.testkit.engine.EngineExecutionResults;
+import org.junit.platform.testkit.engine.Events;
+
+/**
+ * Integration tests for the behavior of the {@link AutoCloseExtension} to release resources after test execution.
+ *
+ * @since 5.11
+ */
+class AutoCloseTests extends AbstractJupiterTestEngineTests {
+
+	private static final List<String> recorder = new ArrayList<>();
+
+	@BeforeEach
+	void resetRecorder() {
+		recorder.clear();
+	}
+
+	@Test
+	void fieldsAreProperlyClosed() {
+		EngineExecutionResults engineExecutionResults = executeTestsForClass(AutoCloseTestCase.class);
+
+		Events tests = engineExecutionResults.testEvents();
+		tests.assertStatistics(stats -> stats.succeeded(2));
+		// @formatter:off
+		assertEquals(asList(
+				"afterEach-close()", "afterEach-run()",
+				"afterEach-close()", "afterEach-run()",
+					"afterAll-close()"), recorder);
+		// @formatter:onf
```

Comment:
```suggestion
		assertThat(recorder).containsExactly(
				"afterEach-close()",
				"afterEach-run()",
				"afterEach-close()",
				"afterEach-run()",
				"afterAll-close()");
		// @formatter:on
```

For newly introduced functionality, we typically prefer the use of AssertJ assertions.

It would be great if you could switch to AssertJ for assertions in this test class. If  not, I'll likely do it after the work has been merged.